query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryAnalyzer.java

```diff
@@ -0,0 +1,90 @@
+package org.infinispan.query.dsl.embedded.impl;
+
+import java.util.Arrays;
+
+import org.infinispan.objectfilter.SortField;
+import org.infinispan.objectfilter.impl.ql.AggregationFunction;
+import org.infinispan.objectfilter.impl.ql.PropertyPath;
+import org.infinispan.objectfilter.impl.syntax.BooleanExpr;
+import org.infinispan.objectfilter.impl.syntax.IndexedFieldProvider;
+import org.infinispan.objectfilter.impl.syntax.parser.AggregationPropertyPath;
+import org.infinispan.objectfilter.impl.syntax.parser.IckleParsingResult;
+import org.infinispan.objectfilter.impl.syntax.parser.ObjectPropertyHelper;
+
+public final class QueryAnalyzer<TypeMetadata> {
+
+   private final ObjectPropertyHelper<TypeMetadata> propertyHelper;
+
+   public QueryAnalyzer(ObjectPropertyHelper<TypeMetadata> propertyHelper) {
+      this.propertyHelper = propertyHelper;
+   }
+
+   public boolean fullIndexingAggregation(IckleParsingResult<TypeMetadata> parsingResult) {
+      BooleanExpr havingClause = parsingResult.getHavingClause();
+      if (havingClause != null) {
+         return false;
+      }
+
+      PropertyPath[] groupBys = parsingResult.getGroupBy();
+      if (groupBys == null || groupBys.length != 1) {
+         return false;
+      }
+
+      IndexedFieldProvider.FieldIndexingMetadata metadata =
+            propertyHelper.getIndexedFieldProvider().get(parsingResult.getTargetEntityMetadata());
+
+      PropertyPath groupBy = groupBys[0];
+      String[] groupByPath = groupBy.asArrayPath();
+      if (!metadata.isAggregable(groupByPath)) {
+         return false;
+      }
+
+      PropertyPath[] projectedPaths = parsingResult.getProjectedPaths();
+      AggregationPropertyPath aggregation = null;
+      PropertyPath projection = null;
+
+      // we need at least one aggregation and one projection (the latter targeting the group by path)
+      for (PropertyPath projectedPath : projectedPaths) {
+         if (projectedPath instanceof AggregationPropertyPath) {
+            aggregation = (AggregationPropertyPath) projectedPath;
+
+            // only count aggregation function is currently supported by Hibernate Search
+            if (!AggregationFunction.COUNT.equals(aggregation.getAggregationFunction())) {
+               return false;
+            }
+            boolean aggregationTargetsAProperty = propertyHelper.hasProperty(parsingResult.getTargetEntityMetadata(), aggregation.asArrayPath());
+            if (aggregationTargetsAProperty) {
+               if (!metadata.isSearchable(aggregation.asArrayPath())) {
+                  return false;
+               }
+            }
```

Comment:
> the path targeted by the aggregation. I'd expect that select count(code) group by status wouldn't work, since Hibernate Search aggregations count the number of documents, not the number of values... ? See my other comment about this.

If you see the test, you can see that only the entities having a code != null are counted. So it seems to be working.

Reply:
About that:

> the number of aggregations. What if there are two (valid) aggregations, e.g. select status, count(employer), count(code) group by status order by status? It seems you will just erase count(employer)

With this contribution if there are two (valid) aggregations, simply it will be a fallback to the current (non optimal) implementation. I wanted to keep the scope of this issue limited to improving the performances of one aggregation only. But feel free to open an issue as a follow up.

Reply:
> With this contribution if there are two (valid) aggregations, simply the execution will be fall-backed to the current (non optimal) implementation

Ok, I didn't see that. I see a check for `groupBys.length != 1` but not for "there is at most 1 aggregation". I must have missed the test too.

> I wanted to keep the scope of this issue limited to improving the performances of one aggregation only

Sure, makes sense.

> But feel free to open an issue as a follow up

Nope, I'm fine ;)